network/server/tchannelthrift/node/service.go

```diff
@@ -478,6 +485,139 @@ func (s *service) FetchBlocksMetadataRaw(tctx thrift.Context, req *rpc.FetchBloc
 	return result, nil
 }
 
+func (s *service) FetchBlocksMetadataRawV2(tctx thrift.Context, req *rpc.FetchBlocksMetadataRawV2Request) (*rpc.FetchBlocksMetadataRawV2Result_, error) {
+	if s.db.IsOverloaded() {
+		s.metrics.overloadRejected.Inc(1)
+		return nil, tterrors.NewInternalError(errServerIsOverloaded)
+	}
+
+	var err error
+	callStart := s.nowFn()
+	defer func() {
+		if err != nil {
+			s.metrics.fetchBlocksMetadata.ReportSuccess(s.nowFn().Sub(callStart))
+		}
+	}()
+
+	ctx := tchannelthrift.Context(tctx)
+	if req.Limit <= 0 {
+		return nil, nil
+	}
+
+	var opts block.FetchBlocksMetadataOptions
+	if req.IncludeSizes != nil {
+		opts.IncludeSizes = *req.IncludeSizes
+	}
+	if req.IncludeChecksums != nil {
+		opts.IncludeChecksums = *req.IncludeChecksums
+	}
+	if req.IncludeLastRead != nil {
+		opts.IncludeLastRead = *req.IncludeLastRead
+	}
+
+	var (
+		nsID      = s.newID(ctx, req.NameSpace)
+		start     = time.Unix(0, req.RangeStart)
+		end       = time.Unix(0, req.RangeEnd)
+		pageToken = &pt.PageToken{}
+	)
+	err = proto.Unmarshal(req.PageToken, pageToken)
+	if err != nil {
+		return nil, tterrors.NewBadRequestError(errInvalidPageToken)
+	}
+
+	fetchedMetadata, nextShardIndex, err := s.db.FetchBlocksMetadata(
+		ctx, nsID, uint32(req.Shard), start, end, req.Limit, pageToken.ShardIndex, opts)
+	if err != nil {
+		s.metrics.fetchBlocksMetadata.ReportError(s.nowFn().Sub(callStart))
+		return nil, convert.ToRPCError(err)
+	}
+	ctx.RegisterFinalizer(context.FinalizerFn(fetchedMetadata.Close))
+
+	result, err := s.getFetchBlocksMetadataRawV2Result(nextShardIndex, opts, fetchedMetadata)
+	// Finalize pooled datastructures regardless of errors
+	ctx.RegisterFinalizer(context.FinalizerFn(func() {
+		s.finalizeFetchBlocksMetadataRawV2Result(result)
```

Comment:
Should also only register finalizer err is not nil right?

Reply:
I updated the comment / code to be more clear, but even in the error case there are pooled objects that need to be returned